src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1153,13 +1170,13 @@ public StateChange ForkProject(
     {
         var projectId = newProjectState.Id;
 
-        Contract.ThrowIfFalse(ProjectStates.ContainsKey(projectId));
-        var newStateMap = ProjectStates.SetItem(projectId, newProjectState);
+        Contract.ThrowIfFalse(ContainsProject(projectId));
+        var newProjectStates = ProjectStates.Replace(oldProjectState, newProjectState);
 
         newDependencyGraph ??= _dependencyGraph;
 
         var newSolutionState = this.Branch(
```

Comment:
i think i may have missed it when reviewing, but add an Assert in Branch that if is passed newProjectStates that they are soted.